src/crawlee/crawlers/_adaptive_playwright/_adaptive_playwright_crawler.py

```diff
@@ -0,0 +1,456 @@
+            """Helper closure that creates new `RequestHandlerRunResult` and delegates request handling to sub crawler.
+
+            Produces `SubCrawlerRun` that either contains filled `RequestHandlerRunResult` or exception.
+            """
+            try:
+                crawl_result = await self.crawl_one_with(
+                    subcrawler_pipeline=subcrawler_pipeline,
+                    context=context,
+                    timeout=self._request_handler_timeout,
+                    result=RequestHandlerRunResult(key_value_store_getter=self.get_key_value_store),
+                    state=use_state,
+                )
+                return SubCrawlerRun(result=crawl_result)
+            except Exception as e:
+                return SubCrawlerRun(exception=e)
+
+        rendering_type_prediction = self.rendering_type_predictor.predict(context.request.url, context.request.label)
+        should_detect_rendering_type = random() < rendering_type_prediction.detection_probability_recommendation
+
+        if not should_detect_rendering_type:
+            self.log.debug(
+                f'Predicted rendering type {rendering_type_prediction.rendering_type} for {context.request.url}'
+            )
+            if rendering_type_prediction.rendering_type == 'static':
+                context.log.debug(f'Running static request for {context.request.url}')
+                self.predictor_state.track_http_only_request_handler_runs()
+
+                static_run = await _run_subcrawler_pipeline(self._static_context_pipeline)
+                if static_run.result and self.result_checker(static_run.result):
+                    await self._commit_result(result=static_run.result, context=context)
+                    return
+                if static_run.exception:
+                    context.log.exception(
+                        msg=f'Static crawler: failed for {context.request.url}', exc_info=static_run.exception
+                    )
+                else:
+                    context.log.warning(f'Static crawler: returned a suspicious result for {context.request.url}')
+                    self.predictor_state.track_rendering_type_mispredictions()
+
+        context.log.debug(f'Running browser request handler for {context.request.url}')
+
+        if should_detect_rendering_type:
+            # Save copy of global state from `use_state` before it can be mutated by browser crawl.
+            # This copy will be used in the static crawl to make sure they both run with same conditions and to
+            # avoid static crawl to modify the state.
+            # (This static crawl is performed only to evaluate rendering type detection.)
+            kvs = await context.get_key_value_store()
+            default_value = dict[str, JsonSerializable]()
+            old_state: dict[str, JsonSerializable] = await kvs.get_value(BasicCrawler.CRAWLEE_STATE_KEY, default_value)
+            old_state_copy = deepcopy(old_state)
```

Comment:
This may run in parallel - we need to watch out for race conditions. I know that the JS version might not handle them optimally either, but we should try harder here.